scipy/spatial/distance.py

```diff
@@ -1355,37 +1431,17 @@ def pdist(X, metric='euclidean', p=2, w=None, V=None, VI=None):
         VI = _validate_mahalanobis_VI(VI, X, m, n)
 
     if callable(metric):
-        # metrics that expects only doubles:
-        if metric in [braycurtis, canberra, chebyshev, cityblock, correlation,
-                      cosine, euclidean, mahalanobis, minkowski, sqeuclidean,
-                      seuclidean, wminkowski]:
-            X = _convert_to_double(X)
-        # metrics that expects only bools:
-        elif metric in [dice, kulsinski, matching, rogerstanimoto, russellrao,
-                        sokalmichener, sokalsneath, yule]:
-            X = _convert_to_bool(X)
-        # metrics that may receive multiple types:
-        elif metric in [hamming, jaccard]:
-            if X.dtype == bool:
-                X = _convert_to_bool(X)
-            else:
-                X = _convert_to_double(X)
-
-        # metrics that expects multiple args
-        if metric == minkowski:
-            def dfun(u, v):
-                return minkowski(u, v, p)
-        elif metric == wminkowski:
-            def dfun(u, v):
-                return wminkowski(u, v, p, w)
-        elif metric == seuclidean:
-            def dfun(u, v):
-                return seuclidean(u, v, V)
-        elif metric == mahalanobis:
-            def dfun(u, v):
-                return mahalanobis(u, v, VI)
-        else:
-            dfun = metric
+        dfun = metric
+        if p != 2:  # if not default
+            dfun = partial(dfun, p=p)
+        if w is not None:
+            dfun = partial(dfun, w=w)
+        elif V is not None:
+            dfun = partial(dfun, V=V)
+        elif VI is not None:
+            dfun = partial(dfun, VI=VI)
+
+        X = _convert_to_double(X)
```

Comment:
If this is added back and the previous code removed it would probably void #6656. In particular *_custom_notdouble and *_calling_convention tests should fail.